force user input in autocomplete

loonix opened the pull request:
Fixes #174

ghiscoding commented:
Thanks for the PR, however if this option is really only used with the Autocomplete Editor, I would rather move it into the `editorOptions`. 

I try to keep, as much as possible, only options that can be used by all Editors inside the `columnEditor.interface.ts`, else the `editorOptions` is a better fit even though that property is generic.

loonix commented:
That sounds reasonable, I changed it to the `editorOptions` and took it out of the `columnEditor`. I removed the type not sure if you want to add it on the `multipleSelectOption.interface.ts`

ghiscoding commented:
If you want the intellisense to work, could maybe change the `editorOptions` to the following
```ts
editorOptions?: MultipleSelectOption | forceUserInput?: boolean | any;
```

I know that keeping the `any` as a type is not too good, it kind of undo the intellisense feature, maybe someday I should add all possible options for each Editors but it's kinda long to do that. For example the  Date Editor uses [Flatpickr](https://flatpickr.js.org/examples/) and so I would have to go and all every possible options into a separate interface and then modify the `editorOptions` like so
```ts
editorOptions?: MultipleSelectOption | FlatpickrOptions | AutoCompleteOptions;
```
but that is kinda long, I don't have time to do that. Unless you or anyone else have a better way of doing it with TypeScript

loonix commented:
I am happy to give a hand on that, and that would be a lot better since we can track the options straight away (maybe on a next pull request)

ghiscoding commented:
Sure, I created the [version 2.x - Project](https://github.com/ghiscoding/Angular-Slickgrid/projects/1) and added some tasks there. I didn't use much before, so if you have better ways of displaying that, let me know, I think it should be enough to start. I can reference a PR to the task when it opens up

ghiscoding commented on src/app/modules/angular-slickgrid/editors/autoCompleteEditor.ts:
```diff
@@ -69,7 +71,7 @@ export class AutoCompleteEditor implements Editor {
     this.labelName = this.customStructure && this.customStructure.label || 'label';
     this.valueName = this.customStructure && this.customStructure.value || 'value';
-
+    this.forceUserInput = this.columnEditor && this.columnEditor.editorOptions && this.columnEditor.editorOptions.forceUserInput ? this.columnEditor.editorOptions.forceUserInput : false;
```
To simplify this line, you could maybe use a Getter, for example see this [Getter](https://github.com/ghiscoding/Angular-Slickgrid/blob/c4a29417c01de8fdf81a999afa25a517a6c1e5ce/src/app/modules/angular-slickgrid/editors/autoCompleteEditor.ts#L50). It's just a suggestion, I could also accept the code without it too

loonix replied:
Done, makes sense to refactor it  :)

ghiscoding replied:
Oh man I'm really sorry, I was not clear in my comment, can you change it to the following
```ts
get editorOptions() {
    return this.columnEditor && this.columnEditor.editorOptions || {};
  }
```
then you can call your boolean values (and any other editor options) like so
```ts
serializeValue() {
  if (this.editorOptions.forceUserInput) {
...
```

loonix replied:
No problem at all (never used getter before) made a new pr, let me know if it is what you have in mind. Thanks

ghiscoding commented on src/app/modules/angular-slickgrid/editors/autoCompleteEditor.ts:
```diff
@@ -63,13 +65,17 @@ export class AutoCompleteEditor implements Editor {
     this.labelName = this.customStructure && this.customStructure.label || 'label';
     this.valueName = this.customStructure && this.customStructure.value || 'value';
-
+    this.forceUserInput = this.forceUserInputChecker;
```
do you really need this line after using the Getter?

loonix replied:
not after the changes you requested, removed it.

ghiscoding commented:
Thanks looks good now :)

ghiscoding commented:
Will merge now, I was trying to fix the new Circle-CI and Codecov for unit tests coverage. I finally found the correct config. Sweet, I now have full CI (continuous integration) in the project :)

ghiscoding commented:
@loonix 
Sorry for the delay, as you have noticed I wanted to wait for the other PR to be merged. 
Your update is now available under latest version [2.6.0](https://github.com/ghiscoding/Angular-Slickgrid/releases/tag/v2.6.0) on [NPM](https://www.npmjs.com/package/angular-slickgrid)

loonix commented:
No problem, many thanks for telling me. Is there a slack chanel or irc chanel you use to go?

ghiscoding commented:
I don't really use Slack and there's no channel or any kind since I'm alone on this lol
